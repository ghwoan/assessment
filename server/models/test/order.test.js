import db from '../../utils/database.js';
import Order from "../order.js";

describe("SqlUtil Tests", () => {
   beforeAll(async () => {
     //create the table if not exist
    //  await db.query("TRUNCATE TABLE product_category");
   })
   beforeEach(async () => {
     //clear the table data
     await db.query("TRUNCATE Table sales_order");
      // seed with some data
      await db.execute(
         'INSERT INTO sales_order (customer_name, status, category_id, country, created_date) VALUES (?, ?, ?, ?, ?)',
         ['Kivell', 'Accepted', 1, 'United Kingdom', '2022-01-23']
      )
      await db.execute(
         'INSERT INTO sales_order (customer_name, status, category_id, country, created_date) VALUES (?, ?, ?, ?, ?)',
         ['Jardine', 'Accepted', 2, 'Russia', '2022-01-20']
      )
      await db.execute(
         'INSERT INTO sales_order (customer_name, status, category_id, country, created_date) VALUES (?, ?, ?, ?, ?)',
         ['Jardine', 'Processing', 2, 'Russia', '2022-02-09']
      )
      
   
   });
   afterAll(async () => {
      //await db.query("DROP TABLE students");
      await db.query("TRUNCATE Table sales_order");
      db.end();
   })
   
   test("fetchCountryList", async () => {
      var result = await Order.fetchCountryList();
      let rows = result[0];
      // assert
      expect(rows.length).toBe(2);
      expect(rows).toMatchSnapshot();
   });

   test("fetchCustomerList", async () => {
      var result = await Order.fetchCustomerList();
      let rows = result[0];
      // assert
      expect(rows.length).toBe(2);
      expect(rows.map(r=>r.customer_name)).toContain("Jardine");
      expect(rows).toMatchSnapshot();

   });

   
   test("fetchByCondition", async () => {
      let startDate, endDate, customerFilter, statusFilter, categoryFilter, countryFilter;
      var result = await Order.fetchByCondition();
      let rows = result[0];
      // assert
      console.log(rows);
      expect(rows.length).toBe(3);
      expect(rows).toMatchSnapshot();

      
      startDate = "2022-02-01";
      result = await Order.fetchByCondition(startDate);
      rows = result[0];
      expect(rows.length).toBe(1);
      expect(rows[0].customer_name).toMatch("Jardine");
      expect(rows).toMatchSnapshot();

      startDate = undefined;
      endDate = "2022-02-01";
      //endDate, customerFilter, statusFilter, categoryFilter, countryFilter;
      result = await Order.fetchByCondition(startDate, endDate);
      rows = result[0];
      expect(rows.length).toBe(2);
      expect(rows).toMatchSnapshot();


      //test customerFilter
      startDate = undefined;
      endDate = "2022-02-01";
      customerFilter = { list: ["Jardine"], op: "IN" };
      result = await Order.fetchByCondition(startDate, endDate, customerFilter);
      rows = result[0];
      expect(rows.length).toBe(1);
      expect(rows).toMatchSnapshot();

      
      startDate = undefined;
      endDate = undefined;
      customerFilter = { list: ["Jardine"], op: "OUT" };
      result = await Order.fetchByCondition(startDate, endDate, customerFilter);
      rows = result[0];
      expect(rows.length).toBe(1);
      expect(rows[0].customer_name).toMatch("Kivell");
      expect(rows).toMatchSnapshot();
   });

});