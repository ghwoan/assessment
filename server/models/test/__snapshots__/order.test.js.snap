// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SqlUtil Tests fetchByCondition 1`] = `
[
  {
    "category": "Electronics",
    "category_id": 1,
    "country": "United Kingdom",
    "created_date": 2022-01-22T16:00:00.000Z,
    "customer_name": "Kivell",
    "object_id": 1,
    "status": "Accepted",
    "updated_date": null,
  },
  {
    "category": "Furniture",
    "category_id": 2,
    "country": "Russia",
    "created_date": 2022-01-19T16:00:00.000Z,
    "customer_name": "Jardine",
    "object_id": 2,
    "status": "Accepted",
    "updated_date": null,
  },
  {
    "category": "Furniture",
    "category_id": 2,
    "country": "Russia",
    "created_date": 2022-02-08T16:00:00.000Z,
    "customer_name": "Jardine",
    "object_id": 3,
    "status": "Processing",
    "updated_date": null,
  },
]
`;

exports[`SqlUtil Tests fetchByCondition 2`] = `
[
  {
    "category": "Furniture",
    "category_id": 2,
    "country": "Russia",
    "created_date": 2022-02-08T16:00:00.000Z,
    "customer_name": "Jardine",
    "object_id": 3,
    "status": "Processing",
    "updated_date": null,
  },
]
`;

exports[`SqlUtil Tests fetchByCondition 3`] = `
[
  {
    "category": "Electronics",
    "category_id": 1,
    "country": "United Kingdom",
    "created_date": 2022-01-22T16:00:00.000Z,
    "customer_name": "Kivell",
    "object_id": 1,
    "status": "Accepted",
    "updated_date": null,
  },
  {
    "category": "Furniture",
    "category_id": 2,
    "country": "Russia",
    "created_date": 2022-01-19T16:00:00.000Z,
    "customer_name": "Jardine",
    "object_id": 2,
    "status": "Accepted",
    "updated_date": null,
  },
]
`;

exports[`SqlUtil Tests fetchByCondition 4`] = `
[
  {
    "category": "Furniture",
    "category_id": 2,
    "country": "Russia",
    "created_date": 2022-01-19T16:00:00.000Z,
    "customer_name": "Jardine",
    "object_id": 2,
    "status": "Accepted",
    "updated_date": null,
  },
]
`;

exports[`SqlUtil Tests fetchByCondition 5`] = `
[
  {
    "category": "Electronics",
    "category_id": 1,
    "country": "United Kingdom",
    "created_date": 2022-01-22T16:00:00.000Z,
    "customer_name": "Kivell",
    "object_id": 1,
    "status": "Accepted",
    "updated_date": null,
  },
]
`;

exports[`SqlUtil Tests fetchCountryList 1`] = `
[
  {
    "country": "United Kingdom",
  },
  {
    "country": "Russia",
  },
]
`;

exports[`SqlUtil Tests fetchCustomerList 1`] = `
[
  {
    "customer_name": "Kivell",
  },
  {
    "customer_name": "Jardine",
  },
]
`;
